### PR TITLE
Add Python bindings for elektraMerge

### PR DESCRIPTION
**Pull request by atmaxinger**
Add KeySet merging functionality based on the elektraMerge library. The API is loosely based on the C++ kdb merging API. This is my first time with SWIG so I can't guarantee for anything 😂.

This is a stepping stone for [ElktraInitiative/ansible-libelektra#9](https://github.com/ElektraInitiative/ansible-libelektra/issues/9)

## Basics

- [x] Short descriptions of your changes are in the release notes
      (added as entry in `doc/news/_preparation_next_release.md` which
      contains `_(my name)_`)
      **Please always add something to the release notes.**
- [x] Details of what you changed are in commit messages
      (first line should have `module: short statement` syntax)
- [x] References to issues, e.g. `close #X`, are in the commit messages.
- [x] The buildservers are happy. If not, fix **in this order**:
  - [x] add a line in `doc/news/_preparation_next_release.md`
  - [x] reformat the code with `scripts/dev/reformat-all`
  - [x] make all unit tests pass
  - [x] fix all memleaks
- [x] The PR is rebased with current master.

## Checklist

- [x] I added unit tests for my code
- [ ] I fully described what my PR does in the documentation
      (not in the PR description)
- [ ] I fixed all affected documentation
- [x] I added code comments, logging, and assertions as appropriate (see [Coding Guidelines](https://master.libelektra.org/doc/CODING.md))
- [ ] I updated all meta data (e.g. README.md of plugins and [METADATA.ini](https://master.libelektra.org/doc/METADATA.ini))
- [x] I mentioned [every code](/.reuse/dep5) not directly written by me in [reuse syntax](https://reuse.software/)

## Review

- [ ] Documentation is introductory, concise, good to read and describes everything what the PR does
- [ ] Examples are well chosen and understandable
- [ ] Code is conforming to [our Coding Guidelines](https://master.libelektra.org/doc/CODING.md)
- [ ] APIs are conforming to [our Design Guidelines](https://master.libelektra.org/doc/DESIGN.md)
- [ ] Code is consistent to [our Design Decisions](https://master.libelektra.org/doc/decisions)

## Labels

- [ ] Add the "work in progress" label if you do not want the PR to be reviewed yet.
- [ ] Add the "ready to merge" label **if the basics are fulfilled** and no further pushes are planned by you.

### CODE REVIEW

**Review by markus2330 — CHANGES_REQUESTED** [review]
Amazing work! :fireworks: Thanks for also adding tests! :revolving_hearts:

**Comment by markus2330** [comment]
@manuelm Do you have time for a review? Would be great to have your expertise for SWIG&Python here.

File: src/bindings/swig/python/merge.i [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+       self.mergedKeys = None
+    else:
+       self.mergedKeys = kdb.KeySet(mergedKeys)
+
+    self.mergeInformation = informationKey
+
+  def hasConflicts(self):
+    return self.mergedKeys is None or getConflicts(self.mergeInformation.getKey()) > 0
+
+class ConflictStrategyAbort:
```
[the remark this answers is not in the document]
Comment (atmaxinger) [reply src/bindings/swig/python/merge.i:38]: I wasn't sure about the minimum version of Python that we support, as enums first appeared in 3.4. AFAICT the bindings still support 2.7 (?)

[the remark this answers is not in the document]
Comment (markus2330) [reply src/bindings/swig/python/merge.i:38]: Python 2 support we definitely dropped long time ago (except of pythongen, which is obsolete now and waits to be removed, see #2987). Even Debian stable has Python 3.9, so Python 3.4 should be no problem at all.

Please add a check and documentation for which Python we require.

File: src/bindings/swig/python/merge.i [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+%import "kdb.i"
+
+%{
+#include "kdbmerge.h"
+%}
+
+
+ckdb::KeySet * elektraMerge (ckdb::KeySet * our, ckdb::Key * ourRoot, ckdb::KeySet * their, ckdb::Key * theirRoot, ckdb::KeySet * base, ckdb::Key * baseRoot, ckdb::Key * resultKey,
+		       int strategy, ckdb::Key * informationKey);
+int getConflicts (ckdb::Key * informationKey);
```
[the remark this answers is not in the document]
Comment (atmaxinger) [reply src/bindings/swig/python/merge.i:18]: how do I correctly update the symbols.map file?

[the remark this answers is not in the document]
Comment (atmaxinger) [reply src/bindings/swig/python/merge.i]: I guess I figured it out

File: src/bindings/swig/python/merge.i (hunk 1 above)
[the remark this answers is not in the document]
Comment (atmaxinger) [reply src/bindings/swig/python/merge.i]: Where would I add such a check?

File: src/bindings/swig/python/CMakeLists.txt [hunk 3]
```diff
@@ -119,6 +119,34 @@ else ()
 		endif ()
 	endif (${SWIG_VERSION} VERSION_LESS "4.0.0")
 
+	# merge module
+	add_cppheaders (HDR_FILES)
+	set_source_files_properties (merge.i PROPERTIES CPLUSPLUS ON)
+	set_source_files_properties (merge.i PROPERTIES SWIG_FLAGS "-py3;-extranative")
+	if (CMAKE_VERSION VERSION_LESS 3.8)
+		swig_add_module (swig-python-merge python merge.i)
+	else (CMAKE_VERSION VERSION_LESS 3.8)
```
Comment (atmaxinger) [inline src/bindings/swig/python/CMakeLists.txt:128]: @markus2330 @kodebach I just copied that from above. But I'm not sure if my eyes deceive me, but the conditions in the if () and else () are exactly the same !?, so none of that code will ever execute on CMAKE >= 3.8

File: src/bindings/swig/python/merge.i (hunk 1 above)
[the remark this answers is not in the document]
Comment (markus2330) [reply src/bindings/swig/python/merge.i]: In the CMake file, excluding the whole binding on older Python versions.

**Comment by atmaxinger** [comment]
@markus2330 Debian stretch and Ubuntu Xenial now fail because they have a cmake version < 3.8, which does not yet include `swig_add_library`

**Comment by markus2330** [comment]
> @markus2330 Debian stretch and Ubuntu Xenial now fail because they have a cmake version < 3.8, which does not yet include swig_add_library

It is time to say goodbye to them. Can you remove the Docker images, entries in Jenkinsfile etc.?

The required CMake version should be in the top-level CMakeLists.txt.

**Comment by atmaxinger** [comment]
@markus2330 I changed the minimum required cmake version, but I think I can't do much about the Jenkins and Docker files right now, as the Jenkins build server currently has issues with the `docker login` command - every build fails because of that.

**Comment by markus2330** [comment]
Please report build server issues in #160.

@lukashartl do you know about this problem?

**Comment by manuelm** [comment]
> @manuelm Do you have time for a review? Would be great to have your expertise for SWIG&Python here.

Hi. The swig part looks ok to me.

**Comment by atmaxinger** [comment]
@markus2330 I have removed the now defunct docker images, but now the link check fails as an old news file references to one of the Dockerfiles. How should we handle this? I could remove the link but I'm not sure if modifying old news entries is the best idea.

```
Broken Links:
—————————————
/tmp/cirrus-ci-build/doc/news/2019-08-06_0.9.0.md:755:0 warning: invalid link '../../scripts/docker/debian/stretch/Dockerfile'
—————————————
```

**Comment by markus2330** [comment]
jenkins build libelektra please

**Comment by markus2330** [comment]
> How should we handle this?

Good point, seems like we do not have any docu about that. I created #4454

For this case: replace the link with https://git.libelektra.org/blob/799fecd94dcb482ea9338facaf63d39f8e5ab5bd/scripts/docker/debian/stretch/Dockerfile

**Comment by markus2330** [comment]
jenkins build libelektra please

**Comment by atmaxinger** [comment]
jenkins build libelektra please

**Comment by atmaxinger** [comment]
jenkins build libelektra please

**Comment by atmaxinger** [comment]
jenkins build libelektra please

**Comment by atmaxinger** [comment]
jenkins build libelektra please

**Pull request merged**